querqy-core/src/main/java/querqy/trie/RuleExtractor.java

```diff
@@ -0,0 +1,120 @@
+package querqy.trie;
+
+import querqy.CompoundCharSequence;
+import querqy.LowerCaseCharSequence;
+import querqy.trie.model.ExactMatch;
+import querqy.trie.model.LookupState;
+import querqy.trie.model.PrefixMatch;
+import querqy.trie.model.SuffixMatch;
+
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+public class RuleExtractor<T, U> {
```

Comment:
I think the class name should not mention 'rule' as the `querqy.trie` package has no idea about rules. Isn't this more about 'sequence lookup'? (the values could be of any type)